### PR TITLE
Automated bot tests

### PR DESCRIPTION
**Pull request by kazimuth**
# Description of Changes

This allows contributors to comment "bots please" on a PR and have a bot test run against that PR. Currently the tests use Hetzner. Results are uploaded such that only clockworks labs employees can access them.
See https://github.com/clockworklabs/BitCraftBots/pull/3 .

# Expected complexity level and risk

1

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] *Run the automated bot tests manually.*
- [ ] *Run an end-to end test triggered by "bots please"*. This can't be done until we merge this branch :^)

### CODE REVIEW

File: .github/workflows/bottest.yml [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+name: Benchmarks
+
+env:
+  GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  GH_REPO: ${{ github.repository }}
+
+jobs:
+  benchmark:
+    name: run 30-bot bot test
```
Comment (bfops) [inline .github/workflows/bottest.yml:20]: nit: for some reason my brain struggled to parse this - any opinion on "run test (30 bots)" or something?

File: .github/workflows/bottest.yml [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      - name: Install stable toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          components: clippy
+          toolchain: stable
+          target: wasm32-unknown-unknown
+          override: true
+
+      - name: Built the bots
```
Comment (bfops) [inline .github/workflows/bottest.yml:78]: Is this `Built` on purpose?
```suggestion
      - name: Build the bots
```

File: .github/workflows/bottest.yml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+          echo "Let's hope the bots runner is up to date!!!"
+          SPACETIME_BIN="$PWD/target/release/spacetime"
+          pushd ~/BitCraftBots
+          python3 -m scripts.bottest "$SPACETIME_BIN"
+          popd
+          mkdir -p bottest-results
+          cp ~/BitCraftBots/results.zip bottest-results/$RESULTS_NAME.zip
+
+      - name: Upload bot test results to DO spaces
+        uses: shallwefootball/s3-upload-action@master
```
Comment (bfops) [inline .github/workflows/bottest.yml:94]: I never know whether I should feel sketched out by using random GH actions like this (vs using the aws cli or something).
@cloutiertyler do you care?

File: .github/workflows/bottest.yml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        run: echo "PR_REF=$(gh pr view $PR_NUMBER --json headRefName --jq .headRefName)" >>"$GITHUB_ENV"
+
+      - name: Checkout sources
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ env.PR_REF || github.ref }}
+          # if we're on master we want to know what the sha of HEAD~1 is so
+          # that we can compare results from it to HEAD (in the "Fetch markdown
+          # summary PR" step). otherwise, we can use a fully shallow checkout
+          fetch-depth: ${{ env.PR_NUMBER && 1 || 2 }}
```
Comment (bfops) [inline .github/workflows/bottest.yml:67]: dumb question - where does the extra fetched commit actually get used? Is it in the `bottest` script itself?

I'm personally into the idea of just always fetching 2 commits (and keeping the comment), but I don't feel strongly about it.

File: .github/workflows/bottest.yml (hunk 3 above)
Comment (cloutiertyler) [reply .github/workflows/bottest.yml:95]: I mean, I don't love it. If it's open source, maybe we could just make our own copy?

Comment (bfops) [reply .github/workflows/bottest.yml:95]: Made https://github.com/clockworklabs/s3-upload-action

File: .github/workflows/bottest.yml (hunk 4 above)
Comment (kazimuth) [reply .github/workflows/bottest.yml:71]: Nope, this is a remnant from the benchmarks. Removed.

**Review by cloutiertyler — APPROVED** [review]
LGTM

**Pull request merged**